airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converters/abstract_stream_state_converter.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import TYPE_CHECKING, Any, List, MutableMapping, Optional
+
+if TYPE_CHECKING:
+    from airbyte_cdk.sources.streams.concurrent.cursor import CursorField
+
+
+class ConcurrencyCompatibleStateType(Enum):
+    date_range = "date-range"
+
+
+class AbstractStreamStateConverter(ABC):
+    START_KEY = "start"
+    END_KEY = "end"
+
+    def get_concurrent_stream_state(self, cursor_field: Optional["CursorField"], state: MutableMapping[str, Any]) -> Optional[MutableMapping[str, Any]]:
+        if not cursor_field:
+            return None
+        if self.is_state_message_compatible(state):
+            return self.deserialize(state)
+        return self.convert_from_sequential_state(cursor_field, state)
+
+    @abstractmethod
+    def deserialize(self, state: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        """
+        Perform any transformations needed for compatibility with the converter.
+        """
+        ...
+
+    @staticmethod
+    def is_state_message_compatible(state: MutableMapping[str, Any]) -> bool:
+        return bool(state) and state.get("state_type") in [t.value for t in ConcurrencyCompatibleStateType]
+
+    @abstractmethod
+    def convert_from_sequential_state(self, cursor_field: "CursorField", stream_state: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        """
+        Convert the state message to the format required by the ConcurrentCursor.
+
+        e.g.
+        {
+            "state_type": ConcurrencyCompatibleStateType.date_range.value,
+            "metadata": { … },
+            "slices": [
+                {starts: 0, end: 1617030403, finished_processing: true}]
+        }
+        """
+        ...
+
+    @abstractmethod
+    def convert_to_sequential_state(self, cursor_field: "CursorField", stream_state: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        """
+        Convert the state message from the concurrency-compatible format to the stream's original format.
+
+        e.g.
+        { "created": 1617030403 }
+        """
+        ...
+
+    @abstractmethod
+    def increment(self, timestamp: Any) -> Any:
+        """
+        Increment a timestamp by a single unit.
+        """
+        ...
+
+    @abstractmethod
+    def merge_intervals(self, intervals: List[MutableMapping[str, Any]]) -> List[MutableMapping[str, Any]]:
+        """
+        Compute and return a list of merged intervals.
+
+        Intervals may be merged if the start time of the second interval is 1 unit or less (as defined by the
+        `increment` method) than the end time of the first interval.
+        """
+        ...
+
+    @abstractmethod
+    def parse_value(self, value: Any) -> Any:
+        """
+        Parse the value of the cursor field into a comparable value.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def zero_value(self) -> Any:
```

Comment:
nit: in its current form, I think this can either be a string or an integer. Maybe this could be restricted to only being a datetime if the [formatting is done in the convert function](https://github.com/airbytehq/airbyte/pull/33531/files#diff-44999ccc78e8a64a5a79f8352f8aa2a45d8dde6ecaa92d3dff0cc773de104197R95).

Reply:
The formatting actually does happen there, [in the DatetimeStreamStateConverter](https://github.com/airbytehq/airbyte/pull/33531/files#diff-44999ccc78e8a64a5a79f8352f8aa2a45d8dde6ecaa92d3dff0cc773de104197R29).

I made the return type Any for the abstract base class in anticipation of creating a converter that does not use a timestamp (e.g. for file-based sources).